test/proposal.test.ts

```diff
@@ -0,0 +1,249 @@
+  let currentSnapshot: unknown;
+  let gnosisSafeManager: GnosisSafeManager;
+
+  const cowDaoSettings: SafeCreationSettings = {
+    owners: [1, 2, 3, 4, 5].map((i) => "0x".padEnd(42, i.toString())),
+    threshold: 5,
+  };
+  const teamConrollerSettings: SafeCreationSettings = {
+    owners: [6, 7, 8].map((i) => "0x".padEnd(42, i.toString())),
+    threshold: 2,
+  };
+  const virtualTokenCreationSettings: VirtualTokenCreationSettings = {
+    merkleRoot: "0x" + "42".repeat(32),
+    usdcToken: "0x0000" + "42".repeat(17) + "01",
+    gnoToken: "0x0000" + "42".repeat(17) + "02",
+    gnoPrice: "31337",
+    wrappedNativeToken: "0x0000" + "42".repeat(17) + "03",
+    nativeTokenPrice: "42424242",
+  };
+  const settings: DeploymentProposalSettings = {
+    cowDao: cowDaoSettings,
+    teamController: teamConrollerSettings,
+    cowToken: {},
+    virtualCowToken: virtualTokenCreationSettings,
+  };
+
+  before(async function () {
+    await setupDeterministicDeployer(deployer);
+    gnosisSafeManager = await GnosisSafeManager.init(deployer);
+  });
+
+  beforeEach(async function () {
+    // Revert the state of the blockchain to clear deterministically deployed
+    // contracts.
+    currentSnapshot = await hre.network.provider.request({
+      method: "evm_snapshot",
+      params: [],
+    });
+  });
+
+  afterEach(async () => {
+    await hre.network.provider.request({
+      method: "evm_revert",
+      params: [currentSnapshot],
+    });
+  });
+
+  describe("deploys expected contracts", function () {
+    let gnosisDao: Contract;
+    let contracts: Record<keyof FinalAddresses | "virtualCowToken", Contract>;
+
+    before(async function () {
+      const { steps, addresses } = await generateProposal(
+        settings,
+        gnosisSafeManager.getDeploymentAddresses(),
+        hre.ethers,
+      );
+
+      gnosisDao = await (
+        await gnosisSafeManager.newSafe([gnosisDaoOwner.address], 1)
+      ).connect(executor);
+
+      let lastResponse: TransactionResponse;
+      for (const step of steps) {
+        lastResponse = await execSafeTransaction(gnosisDao, step, [
+          gnosisDaoOwner,
+        ]);
+      }
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      expect(lastResponse!).not.to.be.undefined;
+
+      const proxies = await contractsCreatedWithCreateCall(
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        lastResponse!,
+        gnosisSafeManager.createCall.address,
+      );
+      expect(proxies).to.have.length(1);
+
+      contracts = {
+        cowDao: gnosisSafeAt(addresses.cowDao).connect(hre.ethers.provider),
+        investorFundsTarget: gnosisSafeAt(
+          addresses.investorFundsTarget,
+        ).connect(hre.ethers.provider),
+        teamController: gnosisSafeAt(addresses.teamController).connect(
+          hre.ethers.provider,
+        ),
+        cowToken: (await hre.ethers.getContractFactory(ContractName.RealToken))
```

Comment:
I find it easier and shorter to simply write hre.ethers.getContractAt(ContractName.RealToken, addresses.cowToken)